// puer/lib/helper/parser.js
// Generated by CoffeeScript 1.9.1
(function() {
  var Parser, argReg, clean, lReg, sReg;

  lReg = /--([^-][-\w]+)/;

  sReg = /-((?:\w|[^-=\<\>]))(?:\s<([^\s]+)>)?/;

  argReg = /--([^-](?:\w|[^-=\<\>])+)(?:[\s=]\<([^\s]+)\>)?/;

  clean = function(str) {
    return str.trim().replace(/\s{2,}/g, " ");
  };

  Parser = (function() {
    function Parser(config) {
      this.options = [];
    }

    Parser.prototype.on = function(options, description, callback) {
      var opt, ref;
      if (typeof description === "function") {
        ref = ["", description], description = ref[0], callback = ref[1];
      }
      opt = this.inspect(options);
      opt.description = description;
      opt.callback = callback;
      return this.options.push(opt);
    };

    Parser.prototype.inspect = function(options) {
      var cur, parsed;
      cur = {};
      options = clean(options);
      parsed = lReg.exec(options);
      if (parsed != null) {
        options = options.replace(parsed[0], "").trim();
        cur.long = parsed[1];
        if (parsed[2] != null) {
          cur.arg = parsed[2];
        }
      }
      parsed = sReg.exec(options);
      if (parsed != null) {
        cur.short = parsed[1];
        if (parsed[2] != null) {
          cur.arg = parsed[2];
        }
      }
      cur.type = cur.arg != null ? "flag" : "switch";
      return cur;
    };

    Parser.prototype.trigger = function(opt, arg) {
      if (typeof opt.callback === "function") {
        opt.callback(arg);
      }
      return opt.callback = null;
    };

    Parser.prototype.run = function(argv) {
      var i, index, item, len, option, ref, results;
      argv = this._preArgv(argv);
      ref = this.options;
      results = [];
      for (i = 0, len = ref.length; i < len; i++) {
        option = ref[i];
        results.push((function() {
          var j, len1, results1;
          results1 = [];
          for (index = j = 0, len1 = argv.length; j < len1; index = ++j) {
            item = argv[index];
            if (item.long === option.long && (option.long != null)) {
              results1.push(this.trigger(option, item.arg));
            } else if (item.short === option.short && (option.short != null)) {
              this.trigger(option, item.arg);
              continue;
            } else {
              results1.push(void 0);
            }
          }
          return results1;
        }).call(this));
      }
      return results;
    };

    Parser.prototype.output = function() {
      var result;
      result = [];
      this.options.forEach(function(opt) {
        var sep;
        sep = "";
        sep += opt.short ? "  -" + opt.short + "," : "     ";
        sep += opt.long ? "--" + opt.long + (opt.arg ? "<" + opt.arg + ">" : "   ") + "\t" : "  \t";
        sep += (opt.description || "") + "\n";
        return result.push(sep);
      });
      return result.join("");
    };

    Parser.prototype._preArgv = function(argv) {
      var option, result;
      option = argv.join(" ");
      option = option.replace(/(?:\s)-(\w{2,})/g, function(a, b, c) {
        var char, i, num, ref;
        char = [];
        for (num = i = 0, ref = b.length; 0 <= ref ? i < ref : i > ref; num = 0 <= ref ? ++i : --i) {
          char.push(" -" + b[num]);
        }
        return char.join(" ");
      });
      option = option.replace(/\=/g, " ");
      result = [];
      option.replace(/--([^-][\w-]*)(\s[^\s-]+)?|-\w(\s[^\s-]+)?/g, function(a) {
        var cur, seps;
        cur = {};
        seps = a.split(/\s+/);
        if (~seps[0].indexOf("--")) {
          cur.long = seps[0].slice(2);
        } else {
          cur.short = seps[0].slice(1);
        }
        cur.arg = seps[1];
        result.push(cur);
        return "";
      });
      return result;
    };

    return Parser;

  })();

  module.exports = Parser;

}).call(this);
